### PR TITLE
Bulk video optimization start

### PR DESCRIPTION
**Pull request by embarks**
## Context
- we added video optimization buttons to the ppc and now we want to have a button which will optimize all of them

## Summary
- We are working closely in the PPC right now so this is for visibility and might change a lot. I wanted to get a draft out as I'll be out for a couple days. In case questions come up or we decide to approach things differently. 

## Relevant Technical Choices
- Pea pod to discuss a refactor to the prepublish checklist to improve efficiency and developer experience using story triggers, developed originally to support the help center. 
- Because we'll be approaching a new design for the PPC, I am going low-touch on changes for the logic here. Eventually we'll have no need for `getPrepublishErrors` and the big reduce in the checklist tab if the alternative implementation goes well. so, I created a new type called "bulk actions" because it's kind of a separate section of the PPC. 
- I think video optimization will have its own separate state within the PPC eventually with the changes coming down the pipe for showing a "progress" state. This change is JUST to trigger video optimization for all the unoptimized videos. 
- The way I was planning on doing it was adding a `BulkVideoOptimization` component which constructs an array of `processVideo` calls and resolves them via `Promise.all` rather than create a bulk video optimization action.
- We might want to make the "process video" more friendly for bulk actions in the future

### QA

- [ ] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1.

### UAT

- [ ] UAT should use the same steps as above.

This PR can be tested by following these steps:

1.

## Checklist

- [ ] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [ ] I have tested this code to the best of my abilities
- [ ] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [ ] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [ ] This PR contains automated tests (unit, integration, and/or e2e) to verify the code works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [ ] I have added documentation where necessary
- [ ] I have added a matching `Type: XYZ` label to the PR

---

Fixes #

### CODE REVIEW

**Comment by spacedmonkey** [comment]
Are there performance issues with starting multiple transcodes at the same time?

**Comment by embarks** [comment]
> Are there performance issues with starting multiple transcodes at the same time?

@spacedmonkey I pushed some changes up now if you wanted to give it a try. if you add more than one large video and click bulk optimize, it will attempt to transcode them at the same time. I was able to get multiple videos transcoded pretty quickly and it almost seemed to queue them though they are initialized at nearly the same time with `Promise.all`. I didn't notice a significant performance degradation. But I run with a lot of memory.

File: assets/src/edit-story/app/prepublish/guidance/media.js [hunk 1]
```diff
@@ -181,3 +186,40 @@ export function videoElementOptimized(element = {}) {
...
+ */
+export function videoElementsOptimized(story) {
+  // for every page return the video elements
+  const unoptimizedVideos = story.pages.reduce((acc, { elements }) => {
+    const videoElements = elements.filter(({ type }) => type === 'video');
+    return [
+      ...acc,
+      ...videoElements.filter(
+        ({ resource }) =>
+          !resource?.local && !resource?.isOptimized && isLargeVideo(resource)
```
Comment (spacedmonkey) [inline assets/src/edit-story/app/prepublish/guidance/media.js:205]: should also remove any resources where isOptimizing is true?

**Comment by spacedmonkey** [comment]
This PR needed rebase, the lints to pass and all uncommented tests to returned before next review. 

Worth noting I tested this locally with 3 12 minutes long videos. It maxed out my Mac ( 2.3 GHz 8-Core Intel Core i9 with 32GB ). 

Screenshot 2021-06-18 at 15 00 35


Performance here does worry about me. Maybe we should limit that number of trancodes done at once upsteam. There is nothing to stop you from dragging 5 videos into the canvas at the moment.

**Comment by embarks** [comment]
> This PR needed rebase, the lints to pass and all uncommented tests to returned before next review.
> 
> Worth noting I tested this locally with 3 12 minutes long videos. It maxed out my Mac ( 2.3 GHz 8-Core Intel Core i9 with 32GB ).
> 
> Screenshot 2021-06-18 at 15 00 35
> 
> Performance here does worry about me. Maybe we should limit that number of trancodes done at once upsteam. There is nothing to stop you from dragging 5 videos into the canvas at the moment.

@spacedmonkey Yeah, it's a draft and we're working on other stuff that's touching this part so it got a little behind. :) I wanted to try out sequencing the promises to see if that did anything. Sequencing them did improve performance a little bit but they still triggered the transcoding for all of them around the same time.

That's good feedback about the videos you tried. I agree performance will probably be an issue, I think we could address this by having a better `bulkOptimize` option which correctly queues the transcoding rather than just sequencing the promises. I'm not sure how to go about that right now, like you say anybody can trigger this by dragging a bunch of videos to their canvas which, it isn't the purpose of this PR to fix that. It's just using what's available rn.

This one will probably sit for a little bit longer while I help pea pod with our other priorities. I looked into the `optimizeVideo` function and saw it relies on the upload queue but started to feel like I was in the weeds for this PR. I think a limit is probably good. I'm not sure where the best place to put that extra state management would be?

**Comment by embarks** [comment]
@spacedmonkey @swissspidy do you both think these sequenced promises will be sufficient for the 1.9 release? 

Here's what it should look like:
Screen Shot 2021-06-23 at 10 18 29 AM

All the issues got updated to follow this design scheme, and this ticket is blocked by that work. But I'm circling around back to this because otherwise we won't have this option in the new PPC. 

We're working hard to re-arrange the PPC so they are organized by issue type rather than by each page and making some needed improvements, and I was thinking this was good enough to go with so we can finish those visual updates to the PPC and retain all the functionality. 

However, it definitely has some performance and UX implications that we should revisit soon. I'd like to work cross-pod so we have the benefit of your knowledge and context around bulk options for transcoding videos.

I was thinking another way I could address this is I could hook into the isTranscoding state and just do one video at a time. Hold the resource array in state and when one video is done optimizing, move on to the next. That's a bit more work but it would avoid potential perf issues.

Thanks in advance! <3

**Comment by embarks** [comment]
Well, I tried a "one at a time" approach and I'm having trouble hooking into the state which tells me what videos have finished being processed so that I can begin the next video in the queue. `optimizeVideo` returns me the `uploadMedia` promise, which resolves long before the transcoding actually completes. 

`useMediaUploadQueue` only seems to update the item states for videos that are being automatically transcoded, and the IDs inside `processed` state from `useLocalMedia` which reference the current videos being processed by `optimizeVideo` are generated as soon as the video starts being processed and aren't references to the video I sent along for processing.

here's the snippet I was working with:
```
  const startTranscodingQueue = useCallback(() => {
    // get the first failing video and optimize it
    setQueue(unoptimizedVideos);
  }, [unoptimizedVideos]);

  useEffect(() => {
    const process = async () => {
      const [firstVideo] = queue;
      await optimizeVideo({ resource: firstVideo.resource }).finally(() => {
        setQueue((oldQueue) => {
          const newQueue = [...oldQueue];
          newQueue.splice(0, 1);
          return newQueue;
        });
      });
      setOptimizing(firstVideo);
    };

    if (!optimizing && queue.length) {
      process();
    }
  }, [optimizeVideo, optimizing, queue]);

  useEffect(() => {
    if (
      optimizing &&
// `processed` from `useMediaUploadQueue` doesn't work here, and `processed` from `useLocalMedia` returns uuids 
// NOT the resource ids, or the full object for reference
      processed.find(
        (item) => item.resource?.id === optimizing.resource?.id
      )
    ) {
      setOptimizing(null);
    }
  }, [optimizing, processed]);
```
The idea would be to trigger `optimizeVideo` one at a time and when it was finished being processed, start transcoding the next video in the queue.

I know this is all kind of vague, but I appreciate any insight I can get here to finish out this feature and I feel like I'm missing some context. :)

**Comment by swissspidy** [comment]
Oh, I see! Yeah getting that correct state is definitely not as straightforward as it should be. 😞  I mean, ideally should really only need a `for` loop or perhaps a generator function.

I think we can definitely help you here by updating the media upload logic to make it easier to get the state information you need.

But before I dive into this, I'd like to get a better understanding of the problem we're trying to solve here.

When you bulk optimize videos, this adds X new videos to the queue. And as you both said, this would be the same as dragging a bunch of videos on the canvas for upload.

(One could of course argue that multiple simultaneous drag & drop uploads are more rare than a user trying to bulk optimize existing videos)

Now, transcoding is of course resource-intensive, so it's done in separate worker threads so the main thread is not blocked. Are separate threads correctly being instantiated for every video? Or is the main thread blocked and the editor freezes? Is 1 process already maxing out CPU? What's the limit of N parallel processes?

Does running everything in sequence really solve these issues? What about alternative solutions like displaying a warning to the user before they start this process? Should we offer them a way to cancel the process?

**Comment by embarks** [comment]
> Oh, I see! Yeah getting that correct state is definitely not as straightforward as it should be. 😞 I mean, ideally should really only need a `for` loop or perhaps a generator function.
> 
> I think we can definitely help you here by updating the media upload logic to make it easier to get the state information you need.
> 
> But before I dive into this, I'd like to get a better understanding of the problem we're trying to solve here.
> 
> When you bulk optimize videos, this adds X new videos to the queue. And as you both said, this would be the same as dragging a bunch of videos on the canvas for upload.
> 
> (One could of course argue that multiple simultaneous drag & drop uploads are more rare than a user trying to bulk optimize existing videos)

Thanks! Yeah I agree that this is a more common case than simultaneous drag and drops.
 
> Now, transcoding is of course resource-intensive, so it's done in separate worker threads so the main thread is not blocked. Are separate threads correctly being instantiated for every video? Or is the main thread blocked and the editor freezes? Is 1 process already maxing out CPU? What's the limit of N parallel processes?

When I did the sequenced `forEach`, the first `uploadFile` step would be sequenced. Meaning, the first `uploadFile` would complete, and so that promise resolved and the next `uploadFile` begins, and so on. The transcoding work for each individual video would start as soon as their `uploadFile` promises resolved, so multiple transcodes were running after all the `uploadFile` promises completed. 

The editor would lag while multiple transcodes were happening at the same time so it seemed it was running low on resources. I didn't see editor completely freeze up or transcoding failures with 2-4 short videos transcoding simultaneously in Chrome (Firefox is another story, 90% of the time pthread throws an error claiming out of memory on the first video). Note I run a macbook pro for work so the amount of resources I have is probably on average much higher than the average user.

> Does running everything in sequence really solve these issues? 

Running in sequence just prevents multiple transcodes from happening at the same time, so it would be just like if the user optimized them one at a time without having to click. It doesn't solve perf issues with each transcode operation itself. But I guess I haven't been able to try `await` with the transcoding itself, which would probably block the main thread!

> What about alternative solutions like displaying a warning to the user before they start this process? Should we offer them a way to cancel the process?

I think from a UX perspective, a cancel feature would be a good add. As for a warning, maybe we should take the video length and FPS into account, that seems to have a big impact on the resource usage from my perspective, but my experimentation was not very scientific. :D Definitely good to think about with regards to what criteria we should use to evaluate videos prior to optimization.

**Pull request closed without merging**